tao_compiler/mlir/disc/transforms/dot_rewriter.cc

```diff
@@ -190,6 +191,469 @@ struct TransposeFoldingConvert : public OpRewritePattern<mhlo::DotGeneralOp> {
+    std::copy(reordered_rhs_non_contracting_tokens.begin(),
+              reordered_rhs_non_contracting_tokens.end(),
+              std::back_inserter(rhs_reordered_tokens));
+  }
+  // result
+  std::copy(reordered_batching_tokens.begin(), reordered_batching_tokens.end(),
+            std::back_inserter(result_reordered_tokens));
+  std::copy(reordered_lhs_non_contracting_tokens.begin(),
+            reordered_lhs_non_contracting_tokens.end(),
+            std::back_inserter(result_reordered_tokens));
+  std::copy(reordered_rhs_non_contracting_tokens.begin(),
+            reordered_rhs_non_contracting_tokens.end(),
+            std::back_inserter(result_reordered_tokens));
+}
+
+Value EinsumToDotGeneralPattern::processOperand(
+    PatternRewriter& rewriter, Value original_operand, Location loc,
+    SmallVector<char> original_tokens, SmallVector<char> reordered_tokens,
+    llvm::SmallDenseSet<char> non_contracting_tokens,
+    llvm::SmallDenseSet<char> contracting_tokens,
+    llvm::SmallDenseSet<char> batching_tokens, bool isBNC) const {
+  int64_t rank = reordered_tokens.size();
+  auto orig_type = original_operand.getType().cast<RankedTensorType>();
+  auto orig_shape = orig_type.getShape();
+  Value result = nullptr;
+  // If need transpose
+  if (original_tokens != reordered_tokens) {
+    SmallVector<int64_t> permutation;
+    for (char t : reordered_tokens) {
+      auto pos = std::find(original_tokens.begin(), original_tokens.end(), t);
+      permutation.push_back(std::distance(original_tokens.begin(), pos));
+    }
+    auto permutation_attr = DenseIntElementsAttr::get(
+        RankedTensorType::get({rank}, rewriter.getI64Type()), permutation);
+    SmallVector<int64_t> transposed_shape;
+    for (int64_t i : permutation) {
+      transposed_shape.push_back(orig_shape[i]);
+    }
+    auto transposed_type =
+        RankedTensorType::get(transposed_shape, orig_type.getElementType());
+    result = rewriter.create<mhlo::TransposeOp>(
+        loc, transposed_type, original_operand, permutation_attr);
+  }
+  size_t num_contracting_dims = contracting_tokens.size();
+  size_t num_non_contracting_dims = non_contracting_tokens.size();
+  // If a reshape is needed, aka, if num of contracting/non_contracting dims > 1
+  if (num_contracting_dims > 1 || num_non_contracting_dims > 1) {
+    SmallVector<SmallVector<int64_t>> reshape_maps;
+    for (int64_t i = 0; i < batching_tokens.size(); ++i) {
+      SmallVector<int64_t> b({i});
+      reshape_maps.push_back(b);
+    }
+    if (isBNC) {
+      // non_contracting
+      SmallVector<int64_t> n;
+      for (int64_t j = 0; j < non_contracting_tokens.size(); ++j) {
+        n.push_back(batching_tokens.size() + j);
+      }
+      reshape_maps.push_back(n);
+      // contracting
+      SmallVector<int64_t> c;
+      for (int64_t j = 0; j < contracting_tokens.size(); ++j) {
+        c.push_back(batching_tokens.size() + non_contracting_tokens.size() +
+                    j);
+      }
```

Comment:
nit: move the calculation of `c` and `n` out of the `if-else` block.

Reply:
done